File: app/src/main/java/ru/devsp/app/mtgcollections/repository/bound/CardLocalBound.java
package ru.devsp.app.mtgcollections.repository.bound;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import ru.devsp.app.mtgcollections.model.api.ApiResponse;
import ru.devsp.app.mtgcollections.model.api.CardApi;
import ru.devsp.app.mtgcollections.model.objects.Card;
import ru.devsp.app.mtgcollections.tools.AppExecutors;

/**
 * Получение и кэширование карт по id
 * Created by gen on 14.09.2017.
 */

public class CardLocalBound extends NetworkBoundResource<List<Card>, List<Card>> {

    private static final int MAX_CACHE_SIZE = 100;

    private static Map<String, List<Card>> mCache = new HashMap<>();

    private CardApi mCardApi;
    private String mId;

    public CardLocalBound(AppExecutors appExecutors, CardApi cardApi) {
        super(appExecutors);
        mCardApi = cardApi;
    }

    public CardLocalBound setParams(String id) {
        mId = id;
        return this;
    }

    @Override
    protected void saveCallResult(@NonNull List<Card> item) {
        if (!item.isEmpty()) {
            if (mCache.size() == MAX_CACHE_SIZE) {
                mCache.clear();
            }
            mCache.put(mId, item);
        }
    }

    @Override
    protected boolean shouldFetch(@Nullable List<Card> data) {
        return !mCache.containsKey(mId);
    }

    @NonNull
    @Override
    protected LiveData<List<Card>> loadSaved() {
        MutableLiveData<List<Card>> card = new MutableLiveData<>();
        if (mCache.containsKey(mId)) {
            card.setValue(mCache.get(mId));
        } else {
            card.setValue(null);
        }
        return card;
    }

    @NonNull
    @Override
    protected LiveData<ApiResponse<List<Card>>> createCall() {
        return mCardApi.getCard(mId);
    }

}

File: app/src/main/java/ru/devsp/app/mtgcollections/repository/bound/CardLocalBound.kt
package ru.devsp.app.mtgcollections.repository.bound


import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData

import java.util.HashMap

import ru.devsp.app.mtgcollections.model.api.CardApi
import ru.devsp.app.mtgcollections.model.objects.Card
import ru.devsp.app.mtgcollections.model.tools.ApiResponse
import ru.devsp.app.mtgcollections.tools.AppExecutors

/**
 * Получение и кэширование карт по id
 * Created by gen on 14.09.2017.
 */

class CardLocalBound(appExecutors: AppExecutors, private val mCardApi: CardApi) : NetworkBound<List<Card>, List<Card>>(appExecutors) {
    private var lid: String = ""

    fun setParams(id: String): CardLocalBound {
        lid = id
        return this
    }

    override fun saveCallResult(data: List<Card>?) {
        if (data != null && !data.isEmpty()) {
            if (cache.size == MAX_CACHE_SIZE) {
                cache.clear()
            }
            cache[lid] = data
        }
    }

    override fun shouldFetch(data: List<Card>?): Boolean {
        return !cache.containsKey(lid)
    }

    override fun loadSaved(): LiveData<List<Card>> {
        val card = MutableLiveData<List<Card>>()
        if (cache.containsKey(lid)) {
            card.setValue(cache[lid])
        } else {
            card.setValue(null)
        }
        return card
    }

    override fun createCall(): LiveData<ApiResponse<List<Card>>> {
        return mCardApi.getCard(lid)
    }

    companion object {

        private const val MAX_CACHE_SIZE = 100

        private val cache = HashMap<String, List<Card>>()
    }

}